app/main.py

```diff
@@ -0,0 +1,474 @@
+                        # Initialising empty dictionary
+                        new_dict = {}
+
+                        # Convert dictionary items from bytes to strings
+                        for dkey, value in device.items():
+                            new_dict[dkey.decode(
+                                "utf-8")] = value.decode("utf-8")
+                        device = new_dict
+                        logging.debug("get: device found: %s", str(device))
+                        break
+                except WatchError:
+                    logging.warning("get: watcherror, trying again")
+                    continue
+                except redis.ConnectionError:
+                    logging.exception("Connection error")
+                    return "Error: error connectiong to Redis host"
+
+                except Exception:
+                    logging.exception("Unexpected error")
+
+        return device
+
+    def set(self, key, field, value):
+        """Set value if device exists, else first initialize device to DEFAULT."""
+        with self.redis.pipeline() as pipe:
+            while True:
+                try:
+                    pipe.watch(key)
+                    if pipe.exists(key) == 0:
+                        logging.warning("set: no device")
+                        pipe.hmset(key, DEFAULT)
+                        pipe.hmset(key, {field: value})
+                        device = pipe.hgetall(key)
+                        pipe.execute()
+                        logging.info("set: device now: %s", str(device))
+                        break
+
+                    else:
+                        pipe.hmset(key, {field: value})
+                        device = pipe.hgetall(key)
+                        logging.info("set: device found: %s", str(device))
+                        break
+
+                except WatchError:
+                    logging.warning("set: watcherror, trying again")
+                    continue
+                except ConnectionError:
+                    logging.exception("Connection error")
+                except Exception:
+                    logging.exception("Unexpected error occured")
+
+            # Initialising empty dictionary
+            new_dict = {}
+
+            # Convert dictionary items from bytes to strings
+            for dkey, value in device.items():
+                new_dict[dkey.decode("utf-8")] = value.decode("utf-8")
+            device = new_dict
+            logging.info("Set: new device = %s", str(device))
+
+            return device
+
+    def delete(self, key):
+        """Delete a key from Redis client."""
+        try:
+            response = self.redis.delete(key)
+            logging.info("response: %s", str(response))
+
+        except ConnectionError:
+            logging.exception("Connection error")
+        except Exception:
+            logging.exception("Unexpected error occured")
+        return response
```

Comment:
If there's an error from redis then this will emit a log message about that error and then proceed to throw a completely different error since `response` will not have been set.  Probably `UnboundLocalError: local variable 'response' referenced before assignment`

It might be better to let these exceptions bubble up to be handled elsewhere.

Reply:
See master/new version.